testnet/stacks-node/src/nakamoto_node/stackerdb_listener.rs

```diff
@@ -0,0 +1,551 @@
+        }
+    }
+
+    /// Run the StackerDB listener.
+    pub fn run(&mut self) -> Result<(), NakamotoNodeError> {
+        info!("StackerDBListener: Starting up");
+
+        let Some(receiver) = &self.receiver else {
+            return Err(NakamotoNodeError::SigningCoordinatorFailure(
+                "StackerDBListener: Failed to obtain the StackerDB event receiver".into(),
+            ));
+        };
+
+        loop {
+            // was the node asked to stop?
+            if !self.node_keep_running.load(Ordering::SeqCst) {
+                info!("StackerDBListener: received node exit request. Aborting");
+                return Ok(());
+            }
+
+            // was the listener asked to stop?
+            if !self.keep_running.load(Ordering::SeqCst) {
+                info!("StackerDBListener: received listener exit request. Aborting");
+                return Ok(());
+            }
+
+            let event = match receiver.recv_timeout(EVENT_RECEIVER_POLL) {
+                Ok(event) => event,
+                Err(std::sync::mpsc::RecvTimeoutError::Timeout) => {
+                    debug!("StackerDBListener: No StackerDB event received. Checking flags and polling again.");
+                    continue;
+                }
+                Err(std::sync::mpsc::RecvTimeoutError::Disconnected) => {
+                    warn!("StackerDBListener: StackerDB event receiver disconnected");
+                    return Err(NakamotoNodeError::SigningCoordinatorFailure(
+                        "StackerDB event receiver disconnected".into(),
+                    ));
+                }
+            };
+
+            // check to see if this event we got is a signer event
+            let is_signer_event =
+                event.contract_id.name.starts_with(SIGNERS_NAME) && event.contract_id.is_boot();
+
+            if !is_signer_event {
+                debug!("StackerDBListener: Ignoring StackerDB event for non-signer contract"; "contract" => %event.contract_id);
+                continue;
+            }
+
+            let modified_slots = &event.modified_slots.clone();
+
+            let Ok(signer_event) = SignerEvent::<SignerMessageV0>::try_from(event).map_err(|e| {
+                warn!("StackerDBListener: Failure parsing StackerDB event into signer event. Ignoring message."; "err" => ?e);
+            }) else {
+                continue;
+            };
+            let SignerEvent::SignerMessages(signer_set, messages) = signer_event else {
+                debug!("StackerDBListener: Received signer event other than a signer message. Ignoring.");
+                continue;
+            };
+            if signer_set != self.signer_set {
+                debug!(
+                    "StackerDBListener: Received signer event for other reward cycle. Ignoring."
+                );
+                continue;
+            };
+            let slot_ids = modified_slots
+                .iter()
```

Comment:
I would guess the compiler is smart enough to make this optimization, but this could be:

```suggestion
                .into_iter()
```